pages/draft-04/migration-notes.md

```diff
@@ -0,0 +1,76 @@
+| `format`          | `format`          | Validation    | Annotation        | -                                                                                                                                                                      |
+| `dependencies`    | `dependencies`    | Core          | Assertion         | The `dependencies` member values were changed to require an array, eliminating the use of single strings.                                                              |
+| `ref` (legacy)    | `ref` (legacy)    | Core          | Applicator        | -                                                                                                                                                                      |
+| Not present       | `allOf`           | Core          | Applicator        | -                                                                                                                                                                      |
+| Not present       | `anyOff`          | Core          | Applicator        | -                                                                                                                                                                      |
+| Not present       | `definitions`     | Validation    | Reserved Location | -                                                                                                                                                                      |
+| Not present       | `maxProperties`   | Validation    | Assertion         | -                                                                                                                                                                      |
+| Not present       | `minProperties`   | Validation    | Assertion         | -                                                                                                                                                                      |
+| Not present       | `not`             | Core          | Applicator        | -                                                                                                                                                                      |
+| Not present       | `oneOf`           | Core          | Applicator        | -                                                                                                                                                                      |
+
+<Infobox label="Note"> Starting with Draft 4, schema identifiers that use an empty URI "" or a fragment-only URI "#" are no longer allowed.
+
+In Draft 3, these identifiers were considered valid:
+
+```
+id: ""
+id: "#"
+```
+
+However, this format is now prohibited from Draft 4 onwards.
+</Infobox>
```

Comment:
No, it was only used as a reference example.

Reply:
Sorry, I didn't understand your response. The infobox is describing a change to how `id` behaves from draft-03 to draft-04. The table above lists all keyword changes that occurred between draft-03 and draft-04. So, I don't see why `id` wouldn't be included.

Reply:
This is what I was trying to communicate with the infobox.

On Draft 3, it was defined as :
_This attribute defines a URI of a schema that contains the full
representation of this schema. When a validator encounters this
attribute, it SHOULD replace the current schema with the schema
referenced by the value's URI (if known and available) and re-validate
the instance. This URI MAY be relative or absolute, and relative URIs
SHOULD be resolved against the URI of the current schema._
And on Draft 4:
_The value for this keyword MUST be a string, and MUST be a valid URI. This URI MUST be normalized, and SHOULD NOT be an empty fragment (#) or the empty URI._

--
So it was not only about `id`. 

But thank you for pointing out `id`; I checked and noticed I skipped adding it here because it didn't change to `$id` until draft 6.

Reply:
> So it was not only about `id`.

Maybe I'm missing something, but I don't think that's true. You've quoted the definition of `$ref` for draft-03, not `id`. Maybe that's a source of confusion? Both definitions describe the behavior of `id` and nothing more. The draft-03 definition is pretty vague and the draft-04 definition isn't great either, but the only clear difference is that empty URIs and empty fragments are discouraged.

Actually, the spec says "SHOULD", not "MUST". So, technically those empty values aren't disallowed, it's just recommended that schema authors don't use them. **I think that's a good argument for not considering this not a change and leaving it out of the migration page altogether.** Implementations in both drafts should consider those values to effectively be a no-op. The only difference is that draft-04 tells schema authors not to use that particular nonsense value. I'd argue that nothing actually changed in the behavior of `id` from draft-03 to draft-04.